File: 23.合并k个升序链表.cs
/*
 * @lc app=leetcode.cn id=23 lang=csharp
 *
 * [23] 合并K个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution
{
    public ListNode MergeKLists(ListNode[] lists)
    {
        if (lists.Length == 0)
        {
            return null;
        }
        return Merge(lists, 0, lists.Length - 1);
    }

    public ListNode Merge(ListNode[] lists, int l, int r)
    {
        if (l > r) return null;

        if (l == r) return lists[l];
        if (r - 1 == l) return MergeTwoLists(lists[l], lists[r]);

        int mid = (r + l) / 2;
        return MergeTwoLists(Merge(lists, l, mid - 1), Merge(lists, mid, r));
    }


    public ListNode MergeTwoLists(ListNode l, ListNode r)
    {
        if (l == null) return r;
        if (r == null) return l;
        ListNode head = new ListNode();
        ListNode tail = head;
        while (l != null && r != null)
        {
            if (l.val <= r.val)
            {
                tail.next = l;
                l = l.next;
            }
            else
            {
                tail.next = r;
                r = r.next;
            }
            tail = tail.next;
        }

        tail.next = (l != null ? l : r);

        return head.next;
    }
}
// @lc code=end

File: 23.合并k个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并K个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        if (lists.size() == 0) return nullptr;

        auto cmp = [](ListNode *a, ListNode *b) { return a->val > b->val; };
        priority_queue<ListNode *, vector<ListNode *>, decltype(cmp)> pq(cmp);

        for (auto i : lists) {
            if (i == nullptr) continue;
            pq.emplace(i);
        }

        ListNode *dummy = new ListNode();
        ListNode *p = dummy;

        while (!pq.empty()) {
            ListNode *node = pq.top();
            pq.pop();
            p->next = node;
            if (node->next != nullptr) {
                pq.emplace(node->next);
            }
            p = p->next;
        }
        return dummy->next;
    }
};
// @lc code=end
